VCF Compare Script
==================

```{r load_initial_data, echo=FALSE}
library(ggplot2)
library(scales)
library(knitr)
source("../../scripts/analysis/vcf_compare_fcns.R")

dir <- "/Users/Dan/Documents/git/Variant-Caller-Pipeline/data/vcf/"

opts_knit$set(root.dir =dir)
setwd(dir)
args <- sprintf("%s", commandArgs(trailingOnly = TRUE))


pairs <- list()
# args <- c("andersen08_radseq.ws225.vcf.gz", "mmp.vcf.gz")
for (i in 1:(length(args)-1)) {
  pairs<-append(pairs,list(c(args[i], args[i+1])))
}
```

## Files

```{r, echo=FALSE, results="asis"}
print(args)
```

```{r def_functions, echo=FALSE}
# Define Functions for generating stats
number_ticks <- function(n) {function(limits) pretty(limits, n)}



```

```{r global_options, include=FALSE}
opts_chunk$set(fig.width=12, fig.height=8, fig.path=paste0('../../data/reports/',paste0(repl_multiple(args, replace_words),collapse="_"),"/"),
               echo=FALSE, warning=FALSE, message=FALSE)
```

```{r, echo=FALSE}

# Generate Data
pair_results <- lapply(pairs, function(x) { get_pair_results(x[1],x[2]) })
concordance_results <- lapply(pairs, function(x) { get_con_matrix(x[1],x[2]) })

```

```{r ind_reports, echo=FALSE}
#-----------------------------------------------#
# Plot Individual vcf results for SNPs + Indels #
#-----------------------------------------------#

# Bring together total SNP counts
snp_indel <- do.call(rbind.data.frame,sapply(pair_results, function(x) { x["SN"] }))

p_list = list(SNPs="number.of.snps",
              Indels="number.of.indels",
              Number_Of_MultiAllelic_sites="number.of.multiallelic.sites",
              Number_Of_MultiAllelic_SNP_Sites="number.of.multiallelic.snp.sites")

plots <- lapply(names(p_list), function(i) {
  ggplot(snp_indel) + 
    geom_bar(stat="identity", aes_string(x="name", y=p_list[[i]])) +
    geom_text( aes_string(x="name",y=p_list[[i]], label=p_list[[i]], vjust=2), color="white") +
    theme(legend.position="bottom", legend.title=element_blank()) +
    labs(y=i, x="Call Set",title=i) +
    scale_y_continuous(breaks=number_ticks(10), labels=scientific_format()) + 
    facet_grid(. ~ Comparison, space = "free", scales="free", drop=T ) +
    theme(legend.position="top", axis.text.x = element_text(angle = 90, hjust = 1))
})
```

## Individual VCF Results
```{r plots, results='asis', echo=FALSE}
plots[[1]]
plots[[2]]
plots[[3]]
plots[[4]]
```

# Number of SNPs/Strain

```{r PSC, results="asis", echo=FALSE}
# Plot SNP data by Strain #

PSC_data <- do.call(rbind.data.frame,sapply(pair_results, function(x) { x["PSC"] }))

PSC_data <- mutate(PSC_data, total_snps = nRefHom + nNonRefHom + nHets) %.%
            mutate(nonRef_snps = nHets + nNonRefHom) %.%
            filter(nRefHom >0, id != Comparison) %.%
            arrange(Comparison)


# Total SNPs
ggplot(PSC_data) + 
    geom_point(aes(x=sample, y= nonRef_snps, group=id, color=id), size=5) +
    labs(title="SNP count/file", x="Strain Name", y="non-reference Allele") +
    theme(legend.position="top", axis.text.x = element_text(angle = 90, hjust = 1)) +
    facet_grid(. ~ Comparison, drop=T, space="free", scales = "free")
```

# Singletons
```{r, results='asis', echo=FALSE}
ggplot(PSC_data) + 
    geom_point(aes(x=sample, y=nSingletons, group=id, color=id), size=5) +
    labs(title="Singletons", x="Strain Name", y="non-reference Allele") +
    theme(legend.position="top", axis.text.x = element_text(angle = 90, hjust = 1))  +
    facet_grid(. ~ Comparison, drop=T, space="free", scales = "free")
```

## Ind. Sample Concordance

```{r ind_conc, results="asis", echo=FALSE}
con_comb <- do.call("rbind", concordance_results)
con_comb$Concordance <- as.numeric(con_comb$Concordance)

con_comb <- filter(con_comb, Query == Sample) %.%
  group_by(Comparison) %.% 
  mutate(sample_avg_conc=mean(Concordance)) %.%
  ungroup() %.%
  group_by(Sample) %.%
  mutate(sample_count=length(Sample)) %.%
  arrange(Comparison)


# Plot concordance of multiple concordance results
ggplot(con_comb) +
    geom_line(position="identity",aes(y=Concordance, x=con_comb$Sample , color=Comparison, group=Comparison)) + 
    geom_point(position="identity",aes(y=Concordance, x=con_comb$Sample , color=Comparison, group=Comparison)) +
    labs(title="Concordance by Strain", x="Sample Name", y="Concordance (%)") +
    theme(legend.position="top", axis.text.x = element_text(angle = 90, hjust = 1))  +
    facet_grid(. ~ Comparison, drop=T, space="free", scales = "free")

```

## Pairwise Concordance

```{r pairwise_con, results="asis", echo=FALSE}

#Plot
lapply(concordance_results, function(x) { con_chart(x) })

## Save Data!
save(list = ls(all = TRUE), file=paste0('../reports/',make.names(paste0(args,collapse="")),".Rdata",collapse=""))
```
